pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,205 @@
+			Type:              azurev1.ServicePrincipal,
+			AllowedNamespaces: &azurev1.AllowedNamespaces{NamespaceList: []string{defaultCAPINamespace}},
+			ClientID:          string(azureClientID),
+			TenantID:          string(azureTenantID),
+			ClientSecret:      corev1.SecretReference{Name: clusterSecretName, Namespace: defaultCAPINamespace},
+		},
+	}
+
+	if err := r.Create(ctx, azureClusterIdentity); err != nil && !apierrors.IsAlreadyExists(err) {
+		return nil, fmt.Errorf("failed to create Azure Cluster Identity: %w", err)
+	}
+
+	target = &azurev1.AzureCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.Infra.Status.InfrastructureName,
+			Namespace: defaultCAPINamespace,
+			// The ManagedBy Annotation is set so CAPI infra providers ignore the InfraCluster object,
+			// as that's managed externally, in this case by this controller.
+			Annotations: map[string]string{
+				clusterv1.ManagedByAnnotation: managedByAnnotationValueClusterCAPIOperatorInfraClusterController,
+			},
+		},
+
+		Spec: azurev1.AzureClusterSpec{
+			AzureClusterClassSpec: azurev1.AzureClusterClassSpec{
+				Location:         location,
+				AzureEnvironment: "AzurePublicCloud",
+				IdentityRef: &corev1.ObjectReference{
+					Name:      r.Infra.Status.InfrastructureName,
+					Namespace: defaultCAPINamespace,
+					Kind:      "AzureClusterIdentity",
+				},
+			},
+			NetworkSpec: azurev1.NetworkSpec{
+				NodeOutboundLB: &azurev1.LoadBalancerSpec{
+					Name: r.Infra.Status.InfrastructureName,
+					BackendPool: azurev1.BackendPool{
+						Name: r.Infra.Status.InfrastructureName,
+					},
+				},
+				Vnet: azurev1.VnetSpec{
+					Name:          providerSpec.Vnet,
+					ResourceGroup: providerSpec.NetworkResourceGroup,
+				},
+			},
+			ResourceGroup: providerSpec.ResourceGroup,
+			ControlPlaneEndpoint: clusterv1.APIEndpoint{
+				Host: apiURL.Hostname(),
+				Port: int32(port),
+			},
+		},
+	}
+
+	if err := r.Create(ctx, target); err != nil {
+		return nil, fmt.Errorf("failed to create InfraCluster: %w", err)
+	}
+
+	log.Info(fmt.Sprintf("InfraCluster '%s/%s' successfully created", defaultCAPINamespace, r.Infra.Status.InfrastructureName))
+
+	return target, nil
+}
+
+// getAzureMAPIProviderSpec returns a Azure Machine ProviderSpec from the the cluster.
+func getAzureMAPIProviderSpec(ctx context.Context, cl client.Client) (*mapiv1beta1.AzureMachineProviderSpec, error) {
+	rawProviderSpec, err := getRawMAPIProviderSpec(ctx, cl)
+	if err != nil {
+		return nil, fmt.Errorf("unable to obtain MAPI ProviderSpec: %w", err)
+	}
+
+	providerSpec := &mapiv1beta1.AzureMachineProviderSpec{}
+	if err := yaml.Unmarshal(rawProviderSpec, providerSpec); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal MAPI ProviderSpec: %w", err)
+	}
+
+	return providerSpec, nil
+}
+
+// getAzureLocation obtains the Azure Server address.
+func (r *InfraClusterController) getAzureLocation(ctx context.Context) (string, error) {
+	if r.Infra.Spec.PlatformSpec.Azure == nil {
```

Comment:
I would expect the spec to be nil in most cases, do we mean status?

Is the location available from the infra status?

Reply:
Yes, you're right. I've reworked this function (addressing your last comment also).
Location [isn't available from the status unfortunately](https://github.com/openshift/api/blob/3e5de946111c67cb1f2b09ff7da569012d15931d/config/v1/types_infrastructure.go#L538), so this function is necessary.